### ISSUE TITLE
Nested Navigation Example - Browser's back button is changing the address without changing the tab 

### ISSUE THREAD

**Issue by osaxma** [csells/go_router#182]
Issue can be observed in this screen recording: 

https://user-images.githubusercontent.com/46427323/142935109-bfb5375a-ce85-4e1f-8cc4-69e1ec305f7b.mov

**Flutter version:** 2.5.3 (stable)
**go_router version:** 2.2.8

**Comment by csells** [comment]
Hmm... That looks like a regression. @lulupointu any thoughts on this?

**Comment by lulupointu** [comment]
I would be surprised if this ever worked since there is nothing implemented to react to widget changes. This is missing:

```dart
  @override
  void didUpdateWidget(covariant FamilyTabsPage oldWidget) {
    if (oldWidget.index != widget.index && widget.index != _controller.index) {
      _controller.animateTo(
        widget.index,
        duration: const Duration(milliseconds: 300),
      );
    }
    super.didUpdateWidget(oldWidget);
  }
```

Also I don't understand why this is here:
```dart
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _controller.index = widget.index;
  }
```

My theory is that you removed a call to `context.go` with the `context` of `_FamilyTabsPageState` (the `_tap`, I would say you added `BuildContext context` which made it depend on `GoRouter` with the wrong `context`). Therefore, before you had the impression that it worked because `didChangeDependencies` was called.
Or maybe this example never worked :smile: 

In any case `didUpdateWidget` is the right this to use here, `didChangeDependencies` is too often use for the wrong reasons imo, while `didUpdateWidget` is so powerful.

**Comment by csells** [comment]
Nice! Not only did you provide a bug fix but you also found why it used to work and suddenly stopped. Amazing!

@osaxma this fix is part of v2.2.9. can you confirm that it's fixed on your end?

**Comment by osaxma** [comment]
Yes, it's working now!
